k2/csrc/fsa_algo.cu

```diff
@@ -258,38 +259,27 @@ void ArcSort(Fsa &src, Fsa *dest, Array1<int32_t> *arc_map /*= nullptr*/) {
   *dest = tmp;
 }
 
-double ShortestDistance(Fsa &fsa, Array1<int32_t> *arc_map /*= nullptr*/) {
-  ContextPtr &context = fsa.Context();
-  K2_CHECK_EQ(fsa.NumAxes(), 2);
-  K2_CHECK_EQ(context->GetDeviceType(), kCpu);
-  int32_t num_states = fsa.Dim0();
-  k2host::Fsa host_fsa = FsaToHostFsa(fsa);
-  Array1<double> state_weights(context, num_states);
-  std::vector<int32_t> tmp_arc_map;
-  ComputeForwardMaxWeights(host_fsa, state_weights.Data(), &tmp_arc_map);
-  if (arc_map != nullptr) *arc_map = Array1<int32_t>(context, tmp_arc_map);
+double ShortestPath(Fsa &src, Fsa *out,
+                    Array1<int32_t> *best_path_arcs /* = nullptr*/) {
+  if (!src.values.IsValid()) return -std::numeric_limits<double>::infinity();
```

Comment:
Actually, I think usually we can suppose arcs are always valid, so I suggest we don't check this every time.

Reply:
So is it checked only in debug build?

Reply:
it's fine to not check this even in debug build as we have checked `src.values.Context()` in the constructor of `Ragged<T>` (if it is invaid, then `src.shape` is invalid as well, that would usually not happen considering the way we create `Ragged` or `RaggedShape`, we always created row_splits with a context)

The main reason we add `IsValid` in `Array1` is to let us know the `Array1` object is not created with the default constructor (no `context`, compared with empty case which also has `Dim() == 0`). Fow now it's just be useful in `RowIds` things since we leave `row_ids` in `RaggedShapeDim` as invalid as default.

BTW, sometimes it's worthy to check if the Ragged object empty or not (e.g. in Transpose, as we will do `TotSize(1) / dim0`, we need make sure it's not empy, i.e. dim0 >=0). But it's OK here to not check this as you have handled such cases with below code.